src/mbgl/layout/symbol_projection.cpp

```diff
@@ -0,0 +1,218 @@
+        const float x = anchorPos[0] / anchorPos[3];
+        const float y = anchorPos[1] / anchorPos[3];
+        const bool inPaddedViewport = (
+                x >= -clippingBuffer[0] &&
+                x <= clippingBuffer[0] &&
+                y >= -clippingBuffer[1] &&
+                y <= clippingBuffer[1]);
+        return inPaddedViewport && frameHistory.isVisible(placementZoom);
+    }
+
+    void addDynamicAttributes(const Point<float>& anchorPoint, const float angle, const float placementZoom,
+            gl::VertexVector<SymbolDynamicLayoutAttributes::Vertex>& dynamicVertexArray) {
+        auto dynamicVertex = SymbolDynamicLayoutAttributes::vertex(anchorPoint, angle, placementZoom);
+        dynamicVertexArray.emplace_back(dynamicVertex);
+        dynamicVertexArray.emplace_back(dynamicVertex);
+        dynamicVertexArray.emplace_back(dynamicVertex);
+        dynamicVertexArray.emplace_back(dynamicVertex);
+    }
+
+    void hideGlyphs(size_t numGlyphs, gl::VertexVector<SymbolDynamicLayoutAttributes::Vertex>& dynamicVertexArray) {
+        const Point<float> offscreenPoint = { -INFINITY, -INFINITY };
+        for (size_t i = 0; i < numGlyphs; i++) {
+            addDynamicAttributes(offscreenPoint, 0, 25, dynamicVertexArray);
+        }
+    }
+
+    struct PlacedGlyph {
+        PlacedGlyph(Point<float> point_, float angle_) : point(point_), angle(angle_) {}
+        Point<float> point;
+        float angle;
+    };
+
+	optional<PlacedGlyph> placeGlyphAlongLine(const float offsetX, const float lineOffsetX, const float lineOffsetY, const bool flip,
+            Point<float> anchorPoint, const uint16_t anchorSegment, const GeometryCoordinates& line, const mat4& labelPlaneMatrix) {
+
+        const float combinedOffsetX = flip ?
+            offsetX - lineOffsetX :
+            offsetX + lineOffsetX;
+
+        int16_t dir = combinedOffsetX > 0 ? 1 : -1;
+
+        float angle = 0.0;
+        if (flip) {
+            // The label needs to be flipped to keep text upright.
+            // Iterate in the reverse direction.
+            dir *= -1;
+            angle = M_PI;
+        }
+
+        if (dir < 0) angle += M_PI;
+
+        int32_t currentIndex = dir > 0 ? anchorSegment : anchorSegment + 1;
+
+        Point<float> current = anchorPoint;
+        Point<float> prev = anchorPoint;
+        float distanceToPrev = 0.0;
+        float currentSegmentDistance = 0.0;
+        const float absOffsetX = std::abs(combinedOffsetX);
+
+        while (distanceToPrev + currentSegmentDistance <= absOffsetX) {
+            currentIndex += dir;
+
+            // offset does not fit on the projected line
+            if (currentIndex < 0 || currentIndex >= static_cast<int32_t>(line.size())) return {};
+
+            prev = current;
+            current = project(convertPoint<float>(line.at(currentIndex)), labelPlaneMatrix);
```

Comment:
I just benchmarked the macosapp with Instruments and the this projection takes around 6% of the total time of `reprojectLineLabels` so I think it should be ok to leave this out. The difference was much higher in -js.